crates/bevy_scene/src/dynamic_scene.rs

```diff
@@ -348,4 +353,51 @@ mod tests {
             "something is wrong with the this test or the code reloading scenes since the relationship between scene entities is broken"
         );
     }
+
+    // Regression test for https://github.com/bevyengine/bevy/issues/14300
+    // Fails before the fix in https://github.com/bevyengine/bevy/pull/15405
+    #[test]
+    fn no_panic_in_map_entities_after_pending_entity_in_hook() {
+        #[derive(Default, Component, Reflect)]
+        #[reflect(Component)]
+        struct A;
+
+        #[derive(Component, Reflect)]
+        #[reflect(Component, MapEntities)]
+        struct B(pub Entity);
+
+        impl MapEntities for B {
+            fn map_entities<M: EntityMapper>(&mut self, entity_mapper: &mut M) {
+                self.0 = entity_mapper.map_entity(self.0);
+            }
+        }
+
+        let reg = AppTypeRegistry::default();
+        {
+            let mut reg_write = reg.write();
+            reg_write.register::<A>();
+            reg_write.register::<B>();
+        }
+
+        let mut scene_world = World::new();
+        scene_world.insert_resource(reg.clone());
+        scene_world.spawn((B(Entity::PLACEHOLDER), A));
+        let scene = DynamicScene::from_world(&scene_world);
```

Comment:
Would it be possible to have a regression test that doesn't involved DynamicScene? I'm not super clear on the internals, but it looks like it might be possible to reproduce the issue with just A and B, and no DynamicScene?

Reply:
Technically, this can be tested entirely in `bevy_ecs` with `SceneEntityMapper`. I can add that one too :+1:

Reply:
But is the Scene part important? Can't we trigger the panic just with MapEntities + hooks?

Reply:
> Technically, this can be tested entirely in bevy_ecs with SceneEntityMapper. I can add that one too 👍

This would be very helpful to make sure it doesn't regress during the BSN work.